bake/bake.go

```diff
@@ -179,46 +180,82 @@ func dedupMap(ms ...map[string]string) map[string]string {
 	return res
 }
 
+func sliceToMap(env []string) (res map[string]string) {
+	res = make(map[string]string)
+	for _, s := range env {
+		kv := strings.SplitN(s, "=", 2)
+		key := kv[0]
+		switch {
+		case len(kv) == 1:
+			res[key] = ""
+		default:
+			res[key] = kv[1]
+		}
+	}
+	return
+}
+
 func ParseFiles(files []File, defaults map[string]string) (_ *Config, err error) {
 	defer func() {
 		err = formatHCLError(err, files)
 	}()
 
 	var c Config
-	var fs []*hcl.File
+	var composeFiles []File
+	var hclFiles []*hcl.File
+	envs := sliceToMap(os.Environ())
 	for _, f := range files {
-		cfg, isCompose, composeErr := ParseComposeFile(f.Data, f.Name)
-		if isCompose {
-			if composeErr != nil {
-				return nil, composeErr
+		switch filepath.Ext(strings.ToLower(f.Name)) {
+		case ".yml", ".yaml":
+			composeFiles = append(composeFiles, f)
+		case ".json":
+			hf, diags := hclparse.NewParser().ParseJSON(f.Data, f.Name)
+			if diags.HasErrors() {
+				return nil, diags
 			}
-			c = mergeConfig(c, *cfg)
-			c = dedupeConfig(c)
-		}
-		if !isCompose {
-			hf, isHCL, err := ParseHCLFile(f.Data, f.Name)
-			if isHCL {
-				if err != nil {
-					return nil, err
+			hclFiles = append(hclFiles, hf)
+		case ".hcl":
+			hf, diags := hclparse.NewParser().ParseHCL(f.Data, f.Name)
+			if diags.HasErrors() {
+				return nil, diags
+			}
+			hclFiles = append(hclFiles, hf)
+		default:
+			e, err := dotenv.UnmarshalBytes(f.Data)
```

Comment:
Previously, files with an unknown extension were parsed using the HCL parser. Now they will be parsed as env files. @tonistiigi Do you think it could be an issue? Maybe instead we could check for `.env` extension and default to hcl?

Reply:
> Previously, files with an unknown extension were parsed using the HCL parser. Now they will be parsed as env files.

What is the difference?

I would prefer to not change the `-f` behavior and only handle `.env` as a special case when parsing compose.

Reply:
Ah, I think I understand you intentions better now.

Still in the first PR maybe `.env` is enough. Just to have compatibility.

For the `-f` we can maybe first split the files between compose and HCL. Then parse HCL and store all the attributes in the process. Then parse compose(for the second time if needed) and pass the attributes and variables.

If we support `-f` this way it shouldn't be only for some special limited set of HCL files that don't have any blocks but ideally all the vars that would be replaced in HCL would be replaced in Compose yaml as well(and additionally only Compose does `.env` as well).

Reply:
> For the `-f` we can maybe first split the files between compose and HCL. Then parse HCL and store all the attributes in the process. Then parse compose(for the second time if needed) and pass the attributes and variables.
> 
> If we support `-f` this way it shouldn't be only for some special limited set of HCL files that don't have any blocks but ideally all the vars that would be replaced in HCL would be replaced in Compose yaml as well(and additionally only Compose does `.env` as well).

Ah right we also need to store the attributes before passing them to HCL/JSON/Compose files when parsed. I missed that.

I will just add `.env` support for compose in this PR and open another one for `-f`.

Reply:
@crazy-max Thankyou for this PR to fix  load .env file from working dir for compose files ! I was facing the exact issue  that's discussed here and it's resolved. 
Very much looking forward to add another -f file to load another .env data file such that the below command will run or please suggest another option to pass a  custom env variable file. Essentially same as https://github.com/docker/buildx/issues/1135

`docker buildx bake -f docker-compose.yml -f .ghcr.env`